### PR TITLE
ubuntu systemd/legacy init support

### PR DESCRIPTION
**Pull request by kiorky**
This makes salt.service alike compatible with ubuntu>= 15.04

With the systemd switch, the modules are broken for now.

It seems good, im just reviewing it a bit more

cc @thatch45

### CODE REVIEW

**Comment by kiorky** [comment]
the sysv mix is inspired from modules/upstart...

**Comment by kiorky** [comment]
i ve just added some additionnal tests.

**Comment by thatch45** [comment]
So Ubuntu still has sysv? The migration to systemd is not complete?

**Comment by kiorky** [comment]
@thatch45 yes it still has sysv

**Comment by anlutro** [comment]
This applies to Debian 8 as well - it also comes with a systemd-sysv bridge, which allows systemd to manage and run sysv services. Salt has been working fine for me, though, but maybe I'm not testing the right things.

Is it correct to hard-code this sysv legacy integration into the systemd module? It might be better to create a sysv_systemd module for OSes that have both installed.

**Comment by kiorky** [comment]
it is not hardcoded  and it can't be that generic as it is tied to the init implementation.
Please also as it is done in  the upstart module, it feels similar, but it is not.

I think that yes you are not testing the right thing, as clearly (at least on ubuntu 15.04, without the sysv) patch, you wont be able to enable/disable services as it will fail on the activation check.

**Comment by kiorky** [comment]
To be exact, the following sls 
```
{% set s = ['rsyslog'] %}
test:
  service.running:
    - names: {{s}}
    - enable: false

test1:
  service.running:
    - names: {{s}}
    - enable: true
    - require:
        - service: test
```
reults of
```
local:
----------
          ID: test
    Function: service.running
        Name: rsyslog
      Result: True
     Comment: Service rsyslog is already disabled, and is in the desired state
     Started: 10:09:19.591035
    Duration: 403.086 ms
     Changes:
----------
          ID: test1
    Function: service.running
        Name: rsyslog
      Result: False
     Comment: Failed when setting service rsyslog to start at boot, but the service was already running
     Started: 10:09:19.995737
    Duration: 743.762 ms
     Changes:

Summary
------------
Succeeded: 1
Failed:    1
------------
Total states run:     2
```

because of falling
```
[INFO    ] Executing command 'systemctl status rsyslog.service' in directory '/root'
[DEBUG   ] output: * rsyslog.service - LSB: enhanced syslogd
   Loaded: loaded (/etc/init.d/rsyslog)
   Active: active (running) since Mon 2015-04-27 01:18:51 CEST; 1 day 8h ago
     Docs: man:systemd-sysv-generator(8)
 Main PID: 1071 (rsyslogd)
   CGroup: /system.slice/rsyslog.service
           `-1071 /usr/sbin/rsyslogd -n

Apr 27 01:18:51 forestguard systemd[1]: Starting System Logging Service...
Apr 27 01:18:51 forestguard systemd[1]: Started System Logging Service.
[INFO    ] Executing command 'systemctl enable rsyslog.service' in directory '/root'
[ERROR   ] Command 'systemctl enable rsyslog.service' failed with return code: 1

```

**Comment by kiorky** [comment]
(be also sure to run enoughtly systemctl daemon-reload, as it can gives you false positives whenever the service has been activated because of the sysv bridge)

File: salt/modules/systemd.py [hunk 1]
```diff
@@ -158,6 +160,123 @@ def systemctl_reload():
     return retcode == 0
 
 
+def _default_runlevel():
```
Comment (lorengordon) [inline salt/modules/systemd.py]: In the current form, I can think of a few ways that `runlevel` might be unset in the return statement. All of which would mean something was horribly broken, of course, but still... Would we maybe want to define a default runlevel as a parameter? i.e. `def _default_runlevel(runlevel=2):`. Or wrap the return with `if runlevel` and raise an exception if it's unset with a nice error message explaining that their system is broken? ;)

File: salt/modules/systemd.py [hunk 2]
```diff
@@ -158,6 +160,123 @@ def systemctl_reload():
...
+    '''
+    # Try to get the "main" default.  If this fails, throw up our
+    # hands and just guess "2", because things are horribly broken
+    try:
+        with salt.utils.fopen('/etc/init/rc-sysinit.conf') as fp_:
+            for line in fp_:
+                if line.startswith('env DEFAULT_RUNLEVEL'):
+                    runlevel = line.split('=')[-1].strip()
+    except Exception:
+        return '2'
```
Comment (cachedout) [inline salt/modules/systemd.py]: Log here, perhaps?

**Comment by cachedout** [comment]
Hi @kiorky. This looks pretty good. I would like to hold off on getting this merged until we have automated testing for Ubuntu 15.

@rallytime and @jfindlay. Do we have an ETA on that yet? (I know you guys are busy, but this will need to happen fairly soon.)

File: salt/modules/systemd.py (hunk 1 above)
Comment (kiorky) [inline salt/modules/systemd.py]: this code is used flawlessly for years in upstart module, no need.

File: salt/modules/systemd.py (hunk 2 above)
Comment (kiorky) [inline salt/modules/systemd.py]: except you want to pollutate the logs, i dont think we should add a log here as it is a sane default anyway.
If we do, upstart module needs also to be patched.

**Comment by kiorky** [comment]
no problem with that with the hope of non concurrent conflicting commits on systemd module :p.
Nevertheless, Ill be less active this week, and available as always only next monday.

File: salt/modules/systemd.py (hunk 1 above)
Comment (lorengordon) [inline salt/modules/systemd.py]: Ahh, I see, I was only looking at the changes for this commit. Didn't realize that code came from another module. Well, if no one is breaking their systems in ways that cause this code to fail, then it surely will always work just fine.

**Comment by jfindlay** [comment]
@cachedout, @jtand and I can work on adding the new ubuntu to the branches tomorrow.

**Comment by jfindlay** [comment]
The current status with ubuntu 15.04 is that the states in the salt-jenkins repository assume there are packages in the saltstack ppa, which are not there and I haven't had time to investigate further.

**Comment by kiorky** [comment]
you ll have maybe to publish a new package to the ppa after activating the vivid serie to have packages built for vivid.

File: tests/unit/modules/systemd_test.py [hunk 3]
```diff
@@ -44,10 +45,28 @@ def test_get_enabled(self):
         '''
             Test to return a list of all enabled services
         '''
+        def sysv(name):
+            if name in ['d', 'e']:
```
Comment (hvnsweeting) [inline tests/unit/modules/systemd_test.py]: should it be ``return name in ['d', 'e']`` ?

Comment (kiorky) [inline tests/unit/modules/systemd_test.py]: matter of style, equivalent on execution.

**Comment by aboe76** [comment]
Tested this with debian 8.0
needed to add salt/utils/systemd.py and this new module systemd.py to my salt minion on debian but it works perfectly:

It now recognizes ntp as a legacy service and enables it with update-rc.d
```
[INFO    ] Executing command 'systemctl is-enabled ntp.service' in directory '/root'
[DEBUG   ] output: Failed to get unit file state for ntp.service: No such file or directory
[INFO    ] Executing command 'runlevel' in directory '/root'
[DEBUG   ] output: N 5
[INFO    ] Service ntp is already enabled, and is in the desired state
[INFO    ] Completed state [ntp] at time 18:56:12.295470
```
```
----------
          ID: ntpd
    Function: service.running
        Name: ntp
      Result: True
     Comment: Service ntp is already enabled, and is in the desired state
     Started: 18:56:11.548765
    Duration: 746.705 ms
     Changes:
```

**Comment by kiorky** [comment]
@basepi @jfindlay did youd forgot that one ?

**Comment by jfindlay** [comment]
ubuntu-15.04 has been added to both develop and 2015.5 on jenkins.

**Comment by kiorky** [comment]
closed by error

**Comment by kiorky** [comment]
@cachedout @basepi do we rebase those commits up to one now that discussion and review is done ?

**Comment by kiorky** [comment]
@basepi @thatch45 @cachedout someone alive here ? :)

**Comment by joehealy** [comment]
Should this be patched into the debian 8 and ubuntu 15.04 packages?

**Pull request merged**